### DISCUSSION TITLE
Allow templates in changelogUrl

### DISCUSSION THREAD

**Discussion by mueller-ma** [renovatebot/renovate#31723] [Suggest an Idea]
### Tell us more.

As fetching Nuget release notes isn't available yet (see https://github.com/renovatebot/renovate/issues/14128), I created a packageRule to set the changelogUrl for some packages:

````
{
  "matchDatasources": ["nuget"],
  "matchPackageNames": ["Example.**"],
  "changelogUrl": "https://<private proget server>/feeds/<private feed>/{{ packageName }}/{{ newVersion }}/metadata"
}
````

However templating isn't allowed in changelogUrl, Can you allow it for that field?

**Comment by rarkins** [comment]
PR welcome to add this

**Reply by neumantm** [reply]
I'm contemplating implementing this feature.

I guess a new field `changelogUrlTemplate` is preferrable over changing the behavior of the existing field.
What do you think?

Also: Any pointers to where in the code the `changelogUrl` field of the packageRules is applied to the upgrade (`BranchUpgradeConfig`) would be helpful.

**Reply by neumantm** [reply]
After some debugging I found that the relevant application of packageRules is [in the flattenUpdates function](https://github.com/renovatebot/renovate/blob/bc9fc2e3710237287801d3b2c7c7d19ed8b10e08/lib/workers/repository/updates/flatten.ts#L124).
I think this is not a sensible place to evaluate a template, so I'm tending to just have the `changelogUrlTemplate` copied to the `BranchUpgradeConfig` and evaluate it at some later time.
Where do you think is a good place for evaluating the template?
Maybe in [`massageUpdateMetadata`](https://github.com/renovatebot/renovate/blob/bc9fc2e3710237287801d3b2c7c7d19ed8b10e08/lib/workers/repository/update/pr/body/index.ts#L19)?

**Reply by rarkins** [reply]
It's preferred to _not_ add another field. The only thing that's hopefully needed is a `template.compile()` on this field in massageMetadata

**Reply by neumantm** [reply]
See https://github.com/renovatebot/renovate/pull/35380
